### ISSUE TITLE
Unified Form Evaluation System

### ISSUE THREAD

**Issue by yamakii** [yamakii/garmin-performance-analysis#42]
## Overview

Implement a unified form evaluation system using statistical models trained on personal data to eliminate evaluation inconsistencies and provide pace-corrected expectations.

**Project Directory**: `docs/project/2025-10-27_unified_form_evaluation_system/`  
**Planning Document**: [planning.md](../blob/main/docs/project/2025-10-27_unified_form_evaluation_system/planning.md)  
**Priority**: High  
**Estimated Time**: 13-19 hours (2-3 days)

---

### Current Issues

1. **Evaluation Contradictions**
   - `efficiency-section-analyst`: "GCT 258ms is excellent (★★★★☆)"
   - `summary-section-analyst`: "GCT 258ms needs improvement (target <250ms)"
   - Cause: Different evaluation criteria, misuse of prompt examples

2. **No Pace Correction**
   - Same baseline for all paces (e.g., GCT 250ms)
   - Fast Run (5:00/km) GCT 216ms vs Easy Run (7:11/km) GCT 258ms
   - Both treated as same efficiency level (incorrect)
   - Data shows 45ms (21%) variation across paces

3. **Token Waste**
   - Agents generate table data (JSON)
   - Evaluation text alone is sufficient

---

### Core Design

**Personal data-driven, pace-corrected baselines for consistent evaluation**

```
Personal 100+ activities → Train models → Store coefficients → Predict & Evaluate
                                              ↓
                          All agents/workers share → Zero contradictions
```

### Key Features

1. **Statistical Models**
   - GCT: Power regression (`v = c * (GCT)^d`, d < 0)
   - VO/VR: Linear regression (`y = a + b * v`)
   - Robust regression (Huber, RANSAC fallback)

2. **DuckDB Tables**
   - `form_baselines`: Model coefficients (α, d, a, b, RMSE, n_samples)
   - `form_evaluations`: Evaluation results (expected, actual, delta, score, star_rating, needs_improvement, evaluation_text)

3. **MCP Extension**
   - `mcp__garmin-db__get_form_evaluations(activity_id)`: Returns evaluation JSON

4. **Agent Optimization**
   - `efficiency-section-analyst`: Simplified (no table data, 70% token reduction)
   - `summary-section-analyst`: Uses `needs_improvement` flag (no achieved goals)

---

### Phase 1: Form Baseline System Core (4-5h)
- [ ] `tools/form_baseline/trainer.py` - Statistical model training
- [ ] `tools/form_baseline/predictor.py` - Expectation prediction
- [ ] `tools/form_baseline/scorer.py` - Scoring & star rating
- [ ] `tools/form_baseline/utils.py` - Outlier removal, conversions
- [ ] `tools/scripts/train_form_baselines.py` - CLI training script
- [ ] `tools/database/db_writer.py` - Add new tables
- [ ] Unit tests (trainer, predictor, scorer, utils)

### Phase 2: Evaluation Logic Integration (2-3h)
- [ ] `tools/form_baseline/evaluator.py` - Evaluate & store
- [ ] `tools/form_baseline/text_generator.py` - Japanese evaluation text
- [ ] `tools/planner/workflow_planner.py` - Call evaluator
- [ ] Integration tests

### Phase 3: MCP Extension (1-2h)
- [ ] `mcp-server/garmin-db/server.py` - `get_form_evaluations()` tool
- [ ] MCP tool tests

### Phase 4: Agent Integration (2-3h)
- [ ] `.claude/agents/efficiency-section-analyst.md` - Simplify prompt
- [ ] `.claude/agents/summary-section-analyst.md` - Use `needs_improvement`
- [ ] Agent output validation

### Phase 5: Report Generation Extension (2-3h)
- [ ] `tools/reporting/report_generator_worker.py` - Read form_evaluations
- [ ] `tools/reporting/templates/detailed_report.j2` - Table generation logic
- [ ] Report validation

### Phase 6: Testing & Validation (2-3h)
- [ ] Train models: `uv run python tools/scripts/train_form_baselines.py`
- [ ] Evaluate activity: `workflow_planner 2025-10-25`
- [ ] Agent execution: 5 agents parallel
- [ ] Report generation: Verify consistency
- [ ] Performance tests (training <5min, evaluation <1s, token reduction 70%)

---

### Functional
- [ ] All agents/workers use same baselines (form_baselines read)
- [ ] Pace correction accurate (power regression, linear regression)
- [ ] Zero contradictions (efficiency ⇔ summary)
- [ ] form_evaluations table populated
- [ ] `get_form_evaluations()` returns correct JSON
- [ ] Natural Japanese evaluation text

### Quality
- [ ] Monotonicity: GCT model d < 0
- [ ] Accuracy: RMSE (GCT<10ms, VO<1cm, VR<1%)
- [ ] Unit test coverage: 80%+
- [ ] Code quality: Black, Ruff, Mypy pass
- [ ] Pre-commit: All hooks pass

### Performance
- [ ] Training time: <5min (1800+ samples)
- [ ] Evaluation time: <1s per activity
- [ ] MCP response: <100ms
- [ ] Token reduction: ~70% (efficiency-section-analyst)

### Validation
- [ ] 2025-10-25 activity: Zero contradictions
  - GCT 258ms → efficiency: "excellent" = summary: "achieved"
- [ ] Pace correction validation:
  - Fast (5:00/km): GCT expected ~215ms
  - Easy (7:11/km): GCT expected ~260ms
- [ ] Report: Table and evaluation text consistent

---

## Expected Outcomes

1. **Evaluation Consistency**: Zero contradictions across all agents
2. **Data-Driven Evaluation**: Personal baselines, no subjective judgment
3. **Token Optimization**: 70% reduction in efficiency-section-analyst
4. **Maintainability**: Update baselines by retraining (no agent modification)
5. **Extensibility**: Easy to add terrain/grade stratification

---

## Technical Constraints

- DuckDB schema changes (no impact on existing data)
- Python 3.12+ (scikit-learn, pandas, numpy)
- MCP server restart required
- Agent prompt changes (stability priority)

---

## Related Documents

- Planning: [docs/project/2025-10-27_unified_form_evaluation_system/planning.md](../blob/main/docs/project/2025-10-27_unified_form_evaluation_system/planning.md)
- Completion: (TBD after implementation)
